packages/world/src/systems/ProgramSystem.sol

```diff
@@ -1,157 +1,120 @@
 import { updateMachineEnergy } from "../utils/EnergyUtils.sol";
 import { getForceField, isForceFieldFragmentActive } from "../utils/ForceFieldUtils.sol";
 import { notify, AttachProgramNotifData, DetachProgramNotifData } from "../utils/NotifUtils.sol";
 
-import { IProgram } from "../prototypes/IProgram.sol";
-import { IChestProgram } from "../prototypes/IChestProgram.sol";
-import { IForceFieldProgram } from "../prototypes/IForceFieldProgram.sol";
-import { IForceFieldFragmentProgram } from "../prototypes/IForceFieldFragmentProgram.sol";
-import { IDisplayProgram } from "../prototypes/IDisplayProgram.sol";
-import { ISpawnTileProgram } from "../prototypes/ISpawnTileProgram.sol";
-import { IBedProgram } from "../prototypes/IBedProgram.sol";
+import { ObjectTypeId } from "../ObjectTypeId.sol";
+import { ObjectTypes } from "../ObjectTypes.sol";
 import { EntityId } from "../EntityId.sol";
 import { Vec3 } from "../Vec3.sol";
-import { safeCallProgram, callProgramOrRevert } from "../utils/callProgram.sol";
+import { ProgramId } from "../ProgramId.sol";
+import { IAttachProgramHook, IDetachProgramHook, IProgramValidator } from "../ProgramInterfaces.sol";
+import { SAFE_PROGRAM_GAS } from "../Constants.sol";
 
 contract ProgramSystem is System {
-  using WorldResourceIdInstance for ResourceId;
+  function attachProgram(EntityId caller, EntityId target, ProgramId program, bytes calldata extraData) public payable {
+    caller.activate();
+    (, Vec3 targetCoord) = caller.requireConnected(target);
+    target = target.baseEntityId();
 
-  function _requireInterface(address programAddress, bytes4 interfaceId) internal view {
-    require(
-      ERC165Checker.supportsInterface(programAddress, interfaceId),
-      "Program does not implement the required interface"
-    );
-  }
+    // ForceField fragments don't have a position on the grid, so we need to handle them differently
+    // if (objectTypeId == ObjectTypes.ForceFieldFragment) {
+    //   // TODO: figure out proximity checks for fragments
+    // }
 
-  function _callForceFieldProgram(EntityId forceFieldEntityId, EntityId fragmentEntityId, bytes memory data) internal {
-    // We know fragment is active because its forcefield exists, so we can use its program
-    ResourceId fragmentProgram = fragmentEntityId.getProgram();
-    if (fragmentProgram.unwrap() != 0) {
-      callProgramOrRevert(fragmentProgram, data);
-    } else {
-      callProgramOrRevert(forceFieldEntityId.getProgram(), data);
-    }
-  }
+    require(!target.getProgram().exists(), "Existing program must be detached");
 
-  function attachProgram(
-    EntityId callerEntityId,
-    EntityId targetEntityId,
-    ResourceId programSystemId,
-    bytes calldata extraData
-  ) public payable {
-    callerEntityId.activate();
-    (, Vec3 targetCoord) = callerEntityId.requireConnected(targetEntityId);
+    (, bool publicAccess) = Systems._get(program.toResourceId());
+    require(!publicAccess, "Program system must be private");
```

Comment:
I assume this wont block programs having functions that can be called outside the world because it can be a separate program system? eg setPrices and onTransfer are in different systems?

Reply:
private systems can still be called by systems in the same namespace. From mud docs:

> A private System can only be called by accounts that have access. This access can be the result of:
> - Access permission to the namespace in which the System is registered.
> - Access permission specifically to the System.